File: src/main/java/com/fengsheng/phase/DrawPhase.java
package com.fengsheng.phase;

import com.fengsheng.Config;
import com.fengsheng.Fsm;
import com.fengsheng.Player;
import com.fengsheng.ResolveResult;
import org.apache.log4j.Logger;

/**
 * 摸牌阶段
 */
public record DrawPhase(Player player) implements Fsm {
    private static final Logger log = Logger.getLogger(DrawPhase.class);

    @Override
    public ResolveResult resolve() {
        if (!player.isAlive()) {
            return new ResolveResult(new NextTurn(player), true);
        }
        log.info(player + "的回合开始了");
        for (Player p : player.getGame().getPlayers()) {
            p.notifyDrawPhase();
        }
        player.draw(Config.HandCardCountEachTurn);
        return new ResolveResult(new MainPhaseIdle(player), true);
    }

    @Override
    public String toString() {
        return player + "的摸牌阶段";
    }
}

File: src/main/kotlin/phase/DrawPhase.kt
package com.fengsheng.phase

import com.fengsheng.Fsm
import com.fengsheng.Player
import com.fengsheng.ResolveResult
import com.fengsheng.skill.getDrawCardCountEachTurn
import org.apache.logging.log4j.kotlin.logger

/**
 * 摸牌阶段
 */
class DrawPhase(override val whoseTurn: Player) : Fsm {
    override fun resolve(): ResolveResult {
        if (!whoseTurn.alive) {
            return ResolveResult(NextTurn(whoseTurn), true)
        }
        if (!whoseTurn.game!!.isEarly) {
            whoseTurn.coefficientA = (whoseTurn.coefficientA + 1) / 2
            whoseTurn.coefficientB /= 2
        }
        whoseTurn.game!!.turn++
        whoseTurn.game!!.realTurn++
        logger.info("${whoseTurn}的回合开始了")
        for (p in whoseTurn.game!!.players) {
            p!!.notifyDrawPhase()
        }
        whoseTurn.draw(whoseTurn.getDrawCardCountEachTurn())
        return ResolveResult(MainPhaseIdle(whoseTurn), true)
    }

    override fun toString(): String {
        return "${whoseTurn}的摸牌阶段"
    }
}
